pkg/util/proxy_test.go

```diff
@@ -0,0 +1,178 @@
+	actualConfig := proxyConfig.ProxyConfig()
+
+	if len(actualConfig) != len(expectedConfig) {
+		t.Fatal("Expected and actual config length differ")
+	}
+
+	for i, actualValue := range actualConfig {
+		if actualValue != expectedConfig[i] {
+			t.Fatalf("Expected '%s', got '%s'", expectedConfig[i], actualValue)
+		}
+	}
+}
+
+func Test_set_to_environment(t *testing.T) {
+	os.Clearenv()
+	defer os.Clearenv()
+
+	proxyConfig, err := NewProxyConfig("http://foobar.com", "https://snafu.de", "42.42.42.42")
+	if err != nil {
+		t.Fatal("Unexpected error creating proxy config")
+	}
+
+	proxyConfig.SetToEnvironment()
+
+	expectedValue := "http://foobar.com"
+	if os.Getenv("HTTP_PROXY") != expectedValue {
+		t.Fatalf("Unexpected proxy setting. Expected '%s'. Got '%s'", expectedValue, os.Getenv("HTTP_PROXY"))
+	}
+
+	expectedValue = "https://snafu.de"
+	if os.Getenv("HTTPS_PROXY") != expectedValue {
+		t.Fatalf("Unexpected proxy setting. Expected '%s'. Got '%s'", expectedValue, os.Getenv("HTTPS_PROXY"))
+	}
+
+	expectedValue = "localhost,127.0.0.1,172.30.1.1,42.42.42.42"
+	if os.Getenv("NO_PROXY") != expectedValue {
+		t.Fatalf("Unexpected proxy setting. Expected '%s'. Got '%s'", expectedValue, os.Getenv("NO_PROXY"))
+	}
+}
+
+func Test_invalid_http_proxy(t *testing.T) {
+	_, err := NewProxyConfig("foo", "", "")
+	if err == nil {
+		t.Fatal("Expected error, but got none")
+	}
+
+	expectedError := "Proxy URL 'foo' is not valid."
+	if err.Error() != expectedError {
+		t.Fatalf("Unexpected error message, expected '%s', but got '%s'", expectedError, err.Error())
+	}
+
+}
+
+func Test_invalid_https_proxy(t *testing.T) {
+	_, err := NewProxyConfig("", "bar", "")
+	if err == nil {
+		t.Fatal("Expected error, but got none")
+	}
+
+	expectedError := "Proxy URL 'bar' is not valid."
+	if err.Error() != expectedError {
+		t.Fatalf("Unexpected error message, expected '%s', but got '%s'", expectedError, err.Error())
+	}
+
+}
+
+func Test_add_no_proxy(t *testing.T) {
+	proxyConfig, err := NewProxyConfig("http://foobar.com", "https://snafu.de", "42.42.42.42")
+	if err != nil {
+		t.Fatal("Unexpected error creating proxy config")
+	}
+
+	expectedNoProxy := "localhost,127.0.0.1,172.30.1.1,42.42.42.42"
+	if expectedNoProxy != proxyConfig.NoProxy() {
+		t.Fatalf("Unexpected no proxy list, expected '%s', but got '%s'", expectedNoProxy, proxyConfig.NoProxy())
+	}
+
+	proxyConfig.AddNoProxy("snafu.com")
+	expectedNoProxy = "localhost,127.0.0.1,172.30.1.1,42.42.42.42,snafu.com"
+	if expectedNoProxy != proxyConfig.NoProxy() {
+		t.Fatalf("Unexpected no proxy list, expected '%s', but got '%s'", expectedNoProxy, proxyConfig.NoProxy())
+	}
+}
+
+func Test_validate_proxy_url(t *testing.T) {
+	urlList := map[string]bool{
+		"http://foo.com:3128":          true,
+		"htt://foo.com:3128":           false,
+		"http://127.0.0.1:3128":        true,
+		"http://foo:bar@test.com:324":  true,
```

Comment:
We will have to define testcases and actually make them fail appropriately long-term. Querystrings are allowed in the ValidURL check, but not for the proxy purpose. Does [ProxyURL]( https://golang.org/pkg/net/http/#ProxyURL) actually enforce something internally? If so, maybe we do not need these test (library code we do not have to test)?